File: ios/App/App/NativeQuestoesPlugin.swift
import Capacitor
import SwiftUI

@objc(NativeQuestoesPlugin)
public class NativeQuestoesPlugin: CAPPlugin {
    
    private var hostingController: UIHostingController<QuestoesView>?

    @objc func openSession(_ call: CAPPluginCall) {
        let titulo = call.getString("titulo") ?? "Questões"
        let startIndex = call.getInt("startIndex") ?? 0
        let contexto = call.getString("contexto") ?? "pratica"
        let questoesRaw = call.getArray("questoes", JSObject.self) ?? []

        let questoes: [QuestaoItemModel] = questoesRaw.compactMap { obj in
            guard let id = obj["id"] as? String,
                  let enunciado = obj["enunciado"] as? String else { return nil }
            return QuestaoItemModel(
                id: id,
                enunciado: enunciado,
                altA: obj["alt_a"] as? String ?? "",
                altB: obj["alt_b"] as? String ?? "",
                altC: obj["alt_c"] as? String ?? "",
                altD: obj["alt_d"] as? String ?? "",
                altE: obj["alt_e"] as? String ?? "",
                gabaritoOficial: (obj["gabarito_oficial"] as? String ?? "A").trimmingCharacters(in: .whitespacesAndNewlines).uppercased(),
                gabaritoComentado: obj["gabarito_comentado"] as? String ?? "",
                disciplina: obj["disciplina"] as? String ?? "",
                assunto: obj["assunto"] as? String ?? "",
                ano: obj["ano"] as? Int ?? 2024,
                banca: obj["banca"] as? String ?? "",
                orgao: obj["orgao"] as? String ?? ""
            )
        }

        DispatchQueue.main.async { [weak self] in
            guard let self = self, let rootVC = self.bridge?.viewController else {
                call.reject("Unable to find root view controller")
                return
            }

            let questoesView = QuestoesView(
                titulo: titulo,
                questoes: questoes,
                startIndex: startIndex,
                contexto: contexto,
                onQuestaoAnswered: { [weak self] questaoId, alternativa, acertou, tempo in
                    self?.notifyListeners("onQuestaoAnswered", data: [
                        "questaoId": questaoId,
                        "alternativa": alternativa,
                        "acertou": acertou,
                        "tempoSegundos": tempo
                    ])
                },
                onSessionCompleted: { [weak self] total, acertos, erros, tempoTotal in
                    self?.notifyListeners("onSessionCompleted", data: [
                        "total": total,
                        "acertos": acertos,
                        "erros": erros,
                        "tempoTotalSegundos": tempoTotal
                    ])
                },
                onClose: { [weak self] in
                    self?.dismissSession()
                    self?.notifyListeners("onClose", data: [:])
                }
            )

            let hostingVC = UIHostingController(rootView: questoesView)
            hostingVC.modalPresentationStyle = .fullScreen
            self.hostingController = hostingVC
            rootVC.present(hostingVC, animated: true) {
                call.resolve(["success": true])
            }
        }
    }

    @objc func closeSession(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            self?.dismissSession()
            call.resolve(["success": true])
        }
    }

    private func dismissSession() {
        hostingController?.dismiss(animated: true, completion: nil)
        hostingController = nil
    }
}

File: android/app/src/main/java/br/com/app/gpu2675756/gpu0e7509bfb7bde52aef412888bb17a456/NativeQuestoesPlugin.kt
package br.com.app.gpu2675756.gpu0e7509bfb7bde52aef412888bb17a456

import android.content.Intent
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "NativeQuestoesPlugin")
class NativeQuestoesPlugin : Plugin() {

    companion object {
        var instance: NativeQuestoesPlugin? = null
        var currentQuestoesJson: String = "[]"
        var currentSessionTitle: String = "Questões"
        var currentStartIndex: Int = 0
        var currentContexto: String = "pratica"
    }

    override fun load() {
        super.load()
        instance = this
    }

    @PluginMethod
    fun openSession(call: PluginCall) {
        val titulo = call.getString("titulo") ?: "Questões"
        val questoesArray = call.getArray("questoes")?.toString() ?: "[]"
        val startIndex = call.getInt("startIndex", 0) ?: 0
        val contexto = call.getString("contexto") ?: "pratica"

        currentSessionTitle = titulo
        currentQuestoesJson = questoesArray
        currentStartIndex = startIndex
        currentContexto = contexto

        val intent = Intent(context, QuestoesNativeActivity::class.java).apply {
            putExtra("titulo", titulo)
            putExtra("startIndex", startIndex)
            putExtra("contexto", contexto)
            flags = Intent.FLAG_ACTIVITY_NEW_TASK
        }

        context.startActivity(intent)
        call.resolve(JSObject().apply { put("success", true) })
    }

    @PluginMethod
    fun closeSession(call: PluginCall) {
        QuestoesNativeActivity.activeActivity?.finish()
        call.resolve(JSObject().apply { put("success", true) })
    }

    fun emitQuestaoAnswered(questaoId: String, alternativa: String, acertou: Boolean, tempoSegundos: Int) {
        val ret = JSObject().apply {
            put("questaoId", questaoId)
            put("alternativa", alternativa)
            put("acertou", acertou)
            put("tempoSegundos", tempoSegundos)
        }
        notifyListeners("onQuestaoAnswered", ret)
    }

    fun emitSessionCompleted(total: Int, acertos: Int, erros: Int, tempoTotalSegundos: Int) {
        val ret = JSObject().apply {
            put("total", total)
            put("acertos", acertos)
            put("erros", erros)
            put("tempoTotalSegundos", tempoTotalSegundos)
        }
        notifyListeners("onSessionCompleted", ret)
    }

    fun emitClose() {
        notifyListeners("onClose", JSObject())
    }
}
